### PR TITLE
Hacking around at using (and abusing) ThreadLocal to store context information

### PR DESCRIPTION
**Pull request by JonathanGiles**
This is questionable code to make it unnecessary for all of our sync APIs to have a Context parameter. Instead, people could write code such as the following:

```java
try (Context ctx = new Context("Name", "Jonathan").addData("Location", "NZ")) {
    ctx.addData("Employer", "Microsoft");
    clientLibraryMethod();
} catch (Exception e) {
    e.printStackTrace();
}
```

Running the test application, the following is printed to the console:

```
Context before try block: [Context NONE, parent=null]
NOW STARTING TRY BLOCK
In client library method, we have a context: [Context key='Employer', value='Microsoft', parent=[Context key='Location', value='NZ', parent=[Context key='Name', value='Jonathan', parent=null]]]
NOW EXITED TRY BLOCK
Context after try block: [Context NONE, parent=null]
```

I do not know if this will work very reliably inside an app server where security managers are in place, it will need to be tested.

### CODE REVIEW

**Comment by alzimmermsft** [comment]
This should solve the issue in needing to pass in a Context object as a parameter to a service call but it feels that the experience and ease of understanding will be lost due to this magical ThreadLocal-Reflections work.

File: core/azure-core/src/main/java/ThreadLocalTest.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    public static void main(String[] args) {
+        ThreadLocalTest testApp = new ThreadLocalTest();
+        testApp.performTask();
+    }
+
+    public void performTask() {
+        System.out.println("Context before try block: " + getThreadLocalContext());
+
+        System.out.println("NOW STARTING TRY BLOCK");
+        try (Context ctx = new Context("Name", "Jonathan").addData("Location", "NZ")) {
```
Comment (alzimmermsft) [inline core/azure-core/src/main/java/ThreadLocalTest.java]: Does this feel better than passing the context directly?

File: core/azure-core/src/main/java/com/azure/core/util/Context.java [hunk 2]
```diff
@@ -77,4 +95,62 @@ public Context addData(Object key, Object value) {
...
+        }
+        return "[Context key='" + key + "', value='" + value + "', parent=" + parent + "]";
+    }
+
+    public static Context getThreadLocalContext() {
+        Context result = Context.NONE;
+
+        try {
+            if (!reflectionInitiated) {
+                threadLocalsField = Thread.class.getDeclaredField("threadLocals");
```
Comment (alzimmermsft) [inline core/azure-core/src/main/java/com/azure/core/util/Context.java]: As mentioned there are security manager implications with reflections, would the be a fallback if this doesn't work?

In the recent storage API overview we removed context as a parameter in the T client methods but continue to include them in the Response<T> clients. Does that feel like a better compromise?

**Comment by rickle-msft** [comment]
I think I need a little more background before I can review this intelligently.

* I think I'm a little unclear on how we understand the current use cases for the Context. I think from v10 and 11 of Storage, I don't really see much of a use case for customers in the way of defining their own policies and passing a Context to interact with those policies in some way as was the initial description of the type, though we have used Context internally for that purpose. There's been a mention of using it for tracing, but I also don't have a good handle on what that actually looks like or how customers would have to interact with the context to enable it. So what are we actually trying to achieve with the Context at this point? And when do we expect customers will want to interact with it?

* I don't have any experience with using threadlocal storage. If someone has a multithreaded application will that interfere with the ability to use threadlocal contexts correctly?

**Comment by JonathanGiles** [comment]
I should have been clearer - this was meant to gauge gut-feeling and round out the discussion over explicit vs implicit contexts. At present we explicitly have Context arguments in our API, and it is noisy. The code above 'works' (ignoring security managers and multi-threading). Taking those into account (particularly the security manager one as the multi-threading situation is easily resolved), I wanted to get other peoples thoughts.

Context at present is definitely going to be used for tracing. Beyond that, there is not much more planned to use it. It is a way for users to configure a pipeline call on a per-method basis, but we haven't exposed any levers that can be pulled here.

**Comment by rickle-msft** [comment]
I see. Thanks for explaining. I think Context is going to be used very infrequently by customers, especially by sync customers, and therefore if we can make it visible only in need-to-use cases, I'm all for it.

**Comment by JonathanGiles** [comment]
This approach is not a good approach - rejecting it.

**Pull request closed without merging**